### ISSUE TITLE
Can you add video / audio files to powerpoints?

### ISSUE THREAD

**Issue by creedmangrum** [singerla/pptx-automizer#105]
Thanks for your work on this library it is amazing. Wondering, as I can't see it in the docs, can you add video and / or audio files to ppt slides with this?

**Comment by singerla** [comment]
Hi! I'm always glad to hear this is useful to others :) But, I'm sorry, there is currently neither video nor audio supported. You need it? :)

**Comment by creedmangrum** [comment]
Yeah we create these presentations where we are pulling in user generated content to a powerpoint. We pull in images and will need to pull in audio files and video files at some point. It's not an immediate need, but I was just wondering if it was at all on the road map for you, or if it was already possible, or if you think that it will be possible and maybe I could help out adding that functionality when we need it.

**Comment by singerla** [comment]
I'll take a look!

**Comment by MP70** [comment]
@singerla you reminded me, I wrote some initial support for video and audio a good while back, it needs to be expanded a bit from memory but at least supports us not breaking existing media and I /think/ loads external.

Is it something you want?

**Comment by singerla** [comment]
Very much appreciated @MP70!

**Comment by MP70** [comment]
I think basically it's because audio and video are unsupported in mainline. It's a bit more complicated than "an element is an element" sadly we have to update refs, copy files, content track and what not. 

I can try and pick the relevant files out from our fork and make a PR but it will be full of merge conflicts at this point and I just can't really prioritise it right now.

**Comment by creedmangrum** [comment]
@MP70 understood. given that i am able to add the videos/audio to pptx files with just PptxGenJs methods, do you have any ideas on how I could just access the `addMedia` function for a slide in PptxGenJs? Would that be possible? I don't really need to even have it be placed great, my client just wants these dynamically generated pptx files to include the audio/video files if possible. Simple solution is totally fine.

**Comment by MP70** [comment]
That still requires us to support audio/video there :)

**Comment by singerla** [comment]
Hi @MP70 and @creedmangrum! I was trying to figure out a short way to implement media files. It seems to be very similar to SVG, because video and audio are xml children of `<p:pic>`. Please find the results of my research in the latest commit for `feature-add-media-files`, along with `SlideWithMedia.pptx` for testing.

A media file requires three relations: the basic png and twice the file itself, which makes it a bit more complicated.

Still to be done:
- Created relations are not updated properly for `a:audioFile` and `p14:media`, because slide xml is not updated between parsing `mediaTarget` and `audioTarget` on `append()`
- Media files are not copied properly, because of filename prefix "image" instead of "media"
- There's a section `<p:timing>` at the bottom of slide.xml. This can imho be removed without ppt complaining.

**Comment by creedmangrum** [comment]
@singerla this to me looks like a great start, from what I can tell. Please let me know where I can help with testing this out. I have a branch of work that is already set up to add the video / image media properly to a pptx. Happy to test and help support where I can.

**Comment by singerla** [comment]
@creedmangrum please refer to my latest commit. It is now possible to add media shapes (which are basically extendend images) to new slides:

(https://github.com/singerla/pptx-automizer/blob/13e83ea3d44fa990094a2cf3f4bee5aa4b0b7e44/__tests__/add-media.test.ts#L14-L16)

Some known issues:
- It is not possible to add a full slide containing media files. It seems as if it has never worked properly to add an (unmodified) slide with e.g. svg images; the issue persits with audio/video. This means, you can only _add_ shapes to _another_ slide.
- All media files are copied twice at the moment, causing the output files to be much larger than necessary. It is required to skip `copyFiles()` for `p14:media`.

You are welcome to test this with your media files! Happy for feedback!

**Comment by creedmangrum** [comment]
@singerla thank you! I have a few meetings today, but I will try it out as soon as I can. This looks greally great and I am excited to try it out.

**Comment by Palke** [comment]
Hi @singerla I have tested this update (I was looking forward for this functionality) and it seems to work decently nice. Only issue is indeed with the file size, as even the "cover" element is being copied multiple times. Do you feel like this multiplication could be resolved in future? Thanks for the amazing work!

**Comment by creedmangrum** [comment]
@Palke what is the workaround you are using for now?

**Comment by singerla** [comment]
Small update on the file size part of this thread: identical media files are no longer copied more than once. pptx-automizer now keeps a checksum index of ppt/media while it builds the output, so an audio or video file that is used by several shapes — or that is referenced twice by the same shape, which is what media elements do (one …/media relation plus one …/audio or …/video relation) — is stored once and shared by all relations.

In our media test the output went from 9 media files down to 4. @Palke, this should make your post-processing workaround unnecessary — no setting needed, it is always on. Same for images, e.g. a logo on twenty slides.

The remaining limitations from earlier in this thread are unchanged: media on a full slide is still not supported, and media shapes need to be added via addElement.
